### PR TITLE
ci: Integrate `bitcoin-tidy` clang-tidy plugin

### PR DESCRIPTION
**Pull request by fanquake**
Demo of integrating the [bitcoin-tidy](https://github.com/theuni/bitcoin-tidy-plugin), [clang-tidy plugin](https://clang.llvm.org/extra/clang-tidy/) written by theuni into our tidy CI job.

The plugin currently has a single check, `bitcoin-unterminated-logprintf`. This would replace our current Python driven, `git-grep`-based, `.cpp` file only, lint-logs linter.

### CODE REVIEW

**Comment by MarcoFalke** [comment]
If this requires clang-15, I'd prefer to wait until that is backported to Ubuntu Jammy instead of hopping the short-term releases.

**Comment by fanquake** [comment]
> If this requires clang-15, I'd prefer to wait until that is backported to Ubuntu Jammy instead of hopping the short-term releases.

This doesn't actually require LLVM 15, it's just that the `run-clang-tidy` script shipped with LLVM 15 has [support for passing `-load` arguments](https://github.com/llvm/llvm-project/commit/b1f1688e90b22dedc829f5abc9a912f18c034fbc) through to clang-tidy.

I've dropped the Kinetic / LLVM 15 bump, in favour of a different approach; patching `-load` support into the llvm 14 `run-clang-tidy` script.

**Comment by theuni** [comment]
Concept ACK (obviously). Thought dump below.

A few things worth considering when it comes to these plugins:
- They do exactly what you tell them to do - no more and no less.
- Tests are really important
- We may want different checks for certain parts of the code.

Elaborating:

c++ is complicated, so defining compiler rules is also complicated. Taking this code for a trivial example:
```c++
struct A{};
void func()
{
    A foo;
    const A bar;
}
```
In testing, one may find that the following query works to find all declarations of type A:
`varDecl(hasType(asString("struct A")))`
However, this will find only the non-const-qualified `foo` and not `bar`.

Instead, we'd want something more like: `varDecl(hasType(cxxRecordDecl(hasName("A"))))`

I'm mentioning this to try to demonstrate what I meant by "they do exactly what you tell them to". It's quite common (especially when just getting started) to end up with an overly-broad or unnecessarily specific query. The first query would appear to be fine if a `const A` is not used anywhere. The best defense against this from what I can tell is adding a test for each failing corner-case.

So I think we'll likely want some pretty thorough documentation that states exactly what the check is intended to do. Additionally, we'll want tests that verify those intentions. In the case of the plugin here, I have added some loose checks here but nothing automated: https://github.com/theuni/bitcoin-tidy/blob/main/test_desig_init.cpp . An example of a corner-case that wasn't caught at first with this check is included here: nested structs with uninitialized members.

tl;dr: before merge, I think we'll want to add a mini test suite to the plugin repo that allows us to verify that checks are working as defined/intended. Additionally, there will need to be some understanding that checks will likely require refinement over time as new corner cases are exposed.

> - We may want different checks for certain parts of the code.

This isn't important for now, but I suspect that the infrastructure for this will end up being more complicated eventually. The obvious outlier (imo) is libbitcoinkernel, where we may want to add more strict checks that other parts of the code (bitcoin-qt for ex) aren't required to abide by.

**Comment by dergoegge** [comment]
Concept ACK

**Comment by fanquake** [comment]
> Would be good to explain how this is different from https://clang.llvm.org/extra/clang-tidy/checks/cppcoreguidelines/pro-type-member-init.html .

Last I checked, turning that on was essentially unusable, due to false-positives. Will take another look.

**Comment by fanquake** [comment]
Going to close this for now. Opened #27315 in regards to adding `cppcoreguidelines-pro-type-member-init` to clang-tidy.

**Comment by theuni** [comment]
Nice, that one's already done :)

I'll update it and we'll start with that one instead.

**Comment by fanquake** [comment]
Rebased, dropped the no-longer needed load patch. Will swap out for the log lint plugin when it's ready.

**Comment by fanquake** [comment]
Updated to use the new plugin, https://github.com/theuni/bitcoin-tidy-plugin, and new check `bitcoin-unterminated-logprintf`.
Updated the PR description.

cc @theuni @stickies-v @dergoegge @TheCharlatan

**Comment by dergoegge** [comment]
`lint-logs.py` can probably be deleted?

**Comment by fanquake** [comment]
> lint-logs.py can probably be deleted?

Added a commit to drop it.

File: ci/test/01_base_install.sh [hunk 1]
```diff
@@ -75,6 +75,12 @@ if [[ "${RUN_TIDY}" == "true" ]]; then
   git clone --depth=1 https://github.com/include-what-you-use/include-what-you-use -b clang_16 "${DIR_IWYU}"/include-what-you-use
   cmake -B "${DIR_IWYU}"/build/ -G 'Unix Makefiles' -DCMAKE_PREFIX_PATH=/usr/lib/llvm-16 -S "${DIR_IWYU}"/include-what-you-use
   make -C "${DIR_IWYU}"/build/ install "$MAKEJOBS"
+
+  if [ ! -d "${DIR_CORE_TIDY}" ]; then
+    git clone --depth=1 https://github.com/theuni/bitcoin-tidy-plugin "${DIR_CORE_TIDY}"
```
[the remark this answers is not in the document]
Comment (fanquake) [reply ci/test/01_base_install.sh:80]: Yea. I'm not sure exactly what we'll do yet, but I think we could just have a "stable" branch in the tidy-plugin repo.

**Comment by fanquake** [comment]
> Also, CI (tidy) fails

Obviously. Given there is a commit here to make it fail?

File: ci/test/01_base_install.sh (hunk 1 above)
[the remark this answers is not in the document]
Comment (fanquake) [reply ci/test/01_base_install.sh:80]: I don't understand what you mean.

**Comment by dergoegge** [comment]
> Obviously. Given there is a commit here to make it fail?

Yes but it's also failing on a different LogPrint statement, so no need for that commit i guess

**Comment by fanquake** [comment]
> Yes but it's also failing on a different LogPrint statement, so no need for that commit i guess

Yea, assuming it's an actual issue, and not a false positive. If it is, then it means our current linter doesn't work anyways.

**Comment by dergoegge** [comment]
> Yea, assuming it's an actual issue, and not a false positive.

At least for the failure in `./test/util/chainstate.h` it's because the current linter only works on cpp files not headers.

**Comment by fanquake** [comment]
Fixed the non-determinism.
Dropped the test commit.
Fixed the issue in chainstate.h.

> it's because the current linter only works on cpp files not headers.

I guess that's another potential +1 for this approach?

File: ci/test/01_base_install.sh (hunk 1 above)
[the remark this answers is not in the document]
Comment (fanquake) [reply ci/test/01_base_install.sh]: Pinned to a commit here.

File: ci/test/01_base_install.sh [hunk 2]
```diff
@@ -75,6 +75,13 @@ if [[ "${RUN_TIDY}" == "true" ]]; then
   git clone --depth=1 https://github.com/include-what-you-use/include-what-you-use -b clang_16 "${DIR_IWYU}"/include-what-you-use
   cmake -B "${DIR_IWYU}"/build/ -G 'Unix Makefiles' -DCMAKE_PREFIX_PATH=/usr/lib/llvm-16 -S "${DIR_IWYU}"/include-what-you-use
   make -C "${DIR_IWYU}"/build/ install "$MAKEJOBS"
+
+  if [ ! -d "${DIR_CORE_TIDY}" ]; then
+    git clone https://github.com/theuni/bitcoin-tidy-plugin "${DIR_CORE_TIDY}"
+    git -C "${DIR_CORE_TIDY}" checkout 1806ef33ff8b14256d766eb9f616a3528aad4464
```
[the remark this answers is not in the document]
Comment (theuni) [reply ci/test/01_base_install.sh:81]: I'm not at all opposed to this. In fact, yeah, that probably makes more sense. Could always move to a repo if it gets too big/busy here.

**Comment by fanquake** [comment]
> Should this also remove the /* Continued */ code-bloat?

Will add this.

Back to draft while we look at an in-tree approach.

**Comment by theuni** [comment]
Heh, really sorry @fanquake .

I think the out-of-tree approach makes sense for (for example) [plugins where we can create attributes and assign them meaning](https://github.com/theuni/bitcoin-core-clang-plugin/commit/05fb23f21df32df4445951123d281172b12ee61b). In that case, the "atomicity" issue goes away, because the code either respects the attribute or it doesn't.

But yeah, now that @MarcoFalke mentions it, that approach is cumbersome when we have to worry about things getting out of sync. So let's see how it looks to do it in-tree.

@fanquake Want to keep going here, or shall I open a new issue?

I guess my only real question is: if we're going to keep it in-tree does it need a full CMake buildsytem as opposed to a simple Makefile? It's kinda overkill for now, but works well and doesn't appear to complicate the c-i in any way. So I guess I'm inclined to leave it unless anyone objects?

**Comment by fanquake** [comment]
> @fanquake Want to keep going here, or shall I open a new PR?

I can convert this over. I think retaining the current build as-is, is ok.

**Comment by fanquake** [comment]
Reworked. I'm guessing some of our linters could fail against the new lint code? If so, I think we should just exclude it from them. Added some more fixups. Updated the PR description. Not 100% sure about the CI (build) dir handling, but this can also be improved.

Added a commit to drop the `/* Continued */` markers. Can be squashed into the previous commit which drops lint-logs.

File: contrib/devtools/bitcoin-tidy/README [hunk 3]
```diff
@@ -0,0 +1,9 @@
+# Bitcoin Tidy
+
+Example Usage:
+
+```bash
+cmake -S . -B build -DLLVM_DIR=/path/to/lib/cmake/llvm -DCMAKE_BUILD_TYPE=Release
+make -j -C build
```
[the remark this answers is not in the document]
Comment (fanquake) [reply contrib/devtools/bitcoin-tidy/README:7]: Added.

File: contrib/devtools/bitcoin-tidy/logprintf.cpp [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+#include <clang/ASTMatchers/ASTMatchFinder.h>
+
+
+namespace {
+AST_MATCHER(clang::StringLiteral, unterminated) {
+    size_t len = Node.getLength();
+    if(len == 0) {
+        return false;
+    }
+    if(Node.getCodeUnit(len-1) == '\n') {
```
[the remark this answers is not in the document]
Comment (fanquake) [reply contrib/devtools/bitcoin-tidy/logprintf.cpp]: clang-formatted everything using `src/.clang-format`.

File: ci/test/06_script_b.sh [hunk 5]
```diff
@@ -148,9 +148,12 @@ if [ "$RUN_FUNCTIONAL_TESTS" = "true" ]; then
 fi
 
 if [ "${RUN_TIDY}" = "true" ]; then
+  cmake -B "${DIR_CORE_TIDY}"/build -DLLVM_DIR=/usr/lib/llvm-16/cmake -DCMAKE_BUILD_TYPE=Release -S "${BASE_ROOT_DIR}"/contrib/devtools/bitcoin-tidy
+  cmake --build "${DIR_CORE_TIDY}"/build "$MAKEJOBS"
```
[the remark this answers is not in the document]
Comment (fanquake) [reply ci/test/06_script_b.sh:152]: Yea, it's only two files, so very fast, especially compared to the runtime of clang-tidy itself.

**Review by maflcko — APPROVED** [review]
lgtm ACK

File: contrib/devtools/bitcoin-tidy/example_logprintf.cpp [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+        LogPrintf(__VA_ARGS__); \
+    } while (0)
+
+void good_func()
+{
+    LogPrintf("hello world!\n");
+}
+void good_func2()
+{
+    LogPrintf("hello world!...");
```
[the remark this answers is not in the document]
Comment (theuni) [reply contrib/devtools/bitcoin-tidy/example_logprintf.cpp:37]: As you said, this one isn't a big deal but it does highlight the general issue of exceptions.

Generally I see a few ways to handle them:
- Change the code to be unambiguous. In this case it'd be something like adding a `LogPrintf_nonewline()` and using as necessary
- Hard-code exceptions in the plugin. This is basically how the current linter works.
- Try to find a preprocessor guard that applies only to clang-tidy. I can't find one, but I assume I'm just missing it.
- ^^, but define our own. Something like `run-clang-tidy-16 -quiet -extra-arg="-DBITCOIN_TIDY" -load=...`

In this case I think our best options are the first or the last.

[the remark this answers is not in the document]
Comment (theuni) [reply contrib/devtools/bitcoin-tidy/example_logprintf.cpp:37]: >     * Try to find a preprocessor guard that applies only to clang-tidy. I can't find one
>       but I assume I'm just missing it.

Whoops, here it is: https://clang.llvm.org/extra/clang-tidy/#suppressing-undesired-diagnostics

Neat. So the exceptions (if it works as documented) can become:
```
LogPrintf("foo..."); // NOLINT(bitcoin-unterminated-logprintf)
```

Which is nice and self-documenting :)

[the remark this answers is not in the document]
Comment (theuni) [reply contrib/devtools/bitcoin-tidy/example_logprintf.cpp:37]: Pushed a commit here which tests this: https://github.com/theuni/bitcoin-tidy-plugin/commit/cdc007022b6cd40320b7e74ab4cbb8e67a6d6e17

It works as expected.

[the remark this answers is not in the document]
Comment (theuni) [reply contrib/devtools/bitcoin-tidy/example_logprintf.cpp:37]: > You could just add a `/* TIDY IGNORE LOG */` (or similar comment) in the source code and then skip those with the plugin?

Just to address this as I agree it'd be a reasonable solution, there's (currently at least) no access to comments from clang-tidy plugins as they're not part of the AST.

[the remark this answers is not in the document]
Comment (theuni) [reply contrib/devtools/bitcoin-tidy/example_logprintf.cpp:37]: Sorry for all the spam here, one more.

Given the nice opt-out above, I've pushed changes to my repo which eliminate all exceptions, so the plugin rule is now very simply: "every logprintf format string must end in a newline". That will mean we need to opt-out of the current "..." uses with NOLINT, which I think is the correct thing to do.

**Comment by fanquake** [comment]
Pulled in the new changes, will update for exceptions shortly.

File: ci/test/00_setup_env.sh [hunk 7]
```diff
@@ -65,6 +65,7 @@ export BASE_BUILD_DIR=${BASE_BUILD_DIR:-$BASE_SCRATCH_DIR/build}
 # This folder exists only on the ci guest, and on the ci host as a volume.
 export PREVIOUS_RELEASES_DIR=${PREVIOUS_RELEASES_DIR:-$BASE_ROOT_DIR/releases/$HOST}
 export DIR_IWYU="${BASE_SCRATCH_DIR}/iwyu"
+export DIR_CORE_TIDY="${BASE_SCRATCH_DIR}/tidy"
```
[the remark this answers is not in the document]
Comment (fanquake) [reply ci/test/00_setup_env.sh]: Yea. Refactored to remove `DIR_CORE_TIDY` entirely.

**Comment by fanquake** [comment]
Dropped `DIR_CORE_TIDY` and just used `BASE_SCRATCH_DIR` directly.
Added `example_logprintf.cpp` to the lint-format-strings.py exclusion list.

**Comment by theuni** [comment]
> As far as I can tell the only place this is used is in these tests, so I'd suggest just deleting them so that we can enforce the rule sanely.

Actually since they test `LogPrintf_` directly, I suggest just fixing up the newline part so that we no longer support that construction.

```diff
diff --git a/src/test/logging_tests.cpp b/src/test/logging_tests.cpp
index 2699d316da..e448805e69 100644
--- a/src/test/logging_tests.cpp
+++ b/src/test/logging_tests.cpp
@@ -83,10 +83,10 @@ BOOST_AUTO_TEST_CASE(logging_timer)
 BOOST_FIXTURE_TEST_CASE(logging_LogPrintf_, LogSetup)
 {
     LogInstance().m_log_sourcelocations = true;
-    LogPrintf_("fn1", "src1", 1, BCLog::LogFlags::NET, BCLog::Level::Debug, "foo1: %s", "bar1\n");
-    LogPrintf_("fn2", "src2", 2, BCLog::LogFlags::NET, BCLog::Level::None, "foo2: %s", "bar2\n");
-    LogPrintf_("fn3", "src3", 3, BCLog::LogFlags::NONE, BCLog::Level::Debug, "foo3: %s", "bar3\n");
-    LogPrintf_("fn4", "src4", 4, BCLog::LogFlags::NONE, BCLog::Level::None, "foo4: %s", "bar4\n");
+    LogPrintf_("fn1", "src1", 1, BCLog::LogFlags::NET, BCLog::Level::Debug, "foo1: %s\n", "bar1");
+    LogPrintf_("fn2", "src2", 2, BCLog::LogFlags::NET, BCLog::Level::None, "foo2: %s\n", "bar2");
+    LogPrintf_("fn3", "src3", 3, BCLog::LogFlags::NONE, BCLog::Level::Debug, "foo3: %s\n", "bar3");
+    LogPrintf_("fn4", "src4", 4, BCLog::LogFlags::NONE, BCLog::Level::None, "foo4: %s\n", "bar4");
     std::ifstream file{tmp_log_path};
     std::vector<std::string> log_lines;
     for (std::string log; std::getline(file, log);) {
```

File: ci/test/00_setup_env.sh (hunk 7 above)
[the remark this answers is not in the document]
Comment (fanquake) [reply ci/test/00_setup_env.sh]: If you don't do it first I'll cleanup `DIR_IWYU`.

**Comment by fanquake** [comment]
Dropped `DIR_IWYU`.
Updated logging tests with the above suggestion.

**Review by maflcko — APPROVED** [review]

**Comment by fanquake** [comment]
> Also, the commit message of the last commit is wrong?

What would you like to see adjusted?

File: ci/test/01_base_install.sh [hunk 8]
```diff
@@ -72,9 +72,9 @@ if [[ ${USE_MEMORY_SANITIZER} == "true" ]]; then
 fi
 
 if [[ "${RUN_TIDY}" == "true" ]]; then
-  git clone --depth=1 https://github.com/include-what-you-use/include-what-you-use -b clang_16 "${DIR_IWYU}"/include-what-you-use
-  cmake -B "${DIR_IWYU}"/build/ -G 'Unix Makefiles' -DCMAKE_PREFIX_PATH=/usr/lib/llvm-16 -S "${DIR_IWYU}"/include-what-you-use
-  make -C "${DIR_IWYU}"/build/ install "$MAKEJOBS"
+  git clone --depth=1 https://github.com/include-what-you-use/include-what-you-use -b clang_16 "${BASE_SCRATCH_DIR}"/include-what-you-use
+  cmake -B "${BASE_SCRATCH_DIR}"/iwyu-build/ -G 'Unix Makefiles' -DCMAKE_PREFIX_PATH=/usr/lib/llvm-16 -S "${BASE_SCRATCH_DIR}"/include-what-you-use
+  make -C "${BASE_SCRATCH_DIR}"/iwyu-build/ install "$MAKEJOBS"
```
[the remark this answers is not in the document]
Comment (fanquake) [reply ci/test/01_base_install.sh:77]: I think we can save further adjustments to the CI system for a followup PR.

**Comment by fanquake** [comment]
> I think the link to the external repo can be dropped,

Ok.

Also fixed the lint regex.

File: ci/test/01_base_install.sh (hunk 8 above)
[the remark this answers is not in the document]
Comment (fanquake) [reply ci/test/01_base_install.sh:77]: The IWYU commit? The rationale was that you asked me above to make the change, which is the only reason I included it in this PR.

[the remark this answers is not in the document]
Comment (fanquake) [reply ci/test/01_base_install.sh:77]: Ok. Going to implement https://github.com/bitcoin/bitcoin/pull/26296#discussion_r1279032950 and fix all issues.

**Review by maflcko — APPROVED** [review]
ACK 3a727cd7ee02c83efcd57d004e6fca8d8e1bb33b 🗑

<details><summary>Show signature</summary>

Signature:

```
untrusted comment: signature from minisign secret key on empty file; verify via: minisign -Vm "${path_to_any_empty_file}" -P RWTRmVTMeKV5noAMqVlsMugDDCyyTSbA3Re5AkUrhvLVln0tSaFWglOw -x "${path_to_this_whole_four_line_signature_blob}"
RUTRmVTMeKV5npGrKx1nqXCw5zeVHdtdYURB/KlyA/LMFgpNCs+SkW9a8N95d+U4AP1RJMi+krxU1A3Yux4bpwZNLvVBKy0wLgM=
trusted comment: ACK 3a727cd7ee02c83efcd57d004e6fca8d8e1bb33b 🗑
6uebzqjnb7gbD/W0sPwdA9sB69ba9mpiZf0I48jxsqT/O0696vGO6XlfTu0E0KtjlUtVcYHcTJhwzmC2ysqoCg==
```

</details>

**Review by stickies-v — COMMENTED** [review]
Approach ACK 3a727cd7ee02c83efcd57d004e6fca8d8e1bb33b

Did a code review too but both the build stuff as well as the clang-tidy syntax is still very new to me so doesn't mean too much.

Are the bitcoin-tidy tests being run in CI with this PR? And perhaps some tidy-test (in example_logprintf.cpp) for multiline strings would be helpful too?

File: contrib/devtools/bitcoin-tidy/example_logprintf.cpp [hunk 9]
```diff
@@ -0,0 +1,91 @@
+// Copyright (c) 2023 Bitcoin Developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+// Warn about any use of LogPrintf that does not end with a newline.
```
[the remark this answers is not in the document]
Comment (theuni) [reply contrib/devtools/bitcoin-tidy/example_logprintf.cpp:5]: This used to be part of the default build, but now needs to be run manually: `make bitcoin-tidy-tests`

It's probably not a bad idea to have them run by c-i, that way we can see that they're actually catching something.

**Comment by theuni** [comment]
I'm a little confused that there are no `NOLINT`s hooked up, and yet c-i is still passing. Beyond the one that @MarcoFalke pointed out above, surely more are still needed right?

File: src/dbwrapper.cpp [hunk 10]
```diff
@@ -76,7 +76,7 @@ class CBitcoinLevelDBLogger : public leveldb::Logger {
 
                 assert(p <= limit);
                 base[std::min(bufsize - 1, (int)(p - base))] = '\0';
-                LogPrintLevel(BCLog::LEVELDB, BCLog::Level::Debug, "%s", base);
+                LogPrintLevel(BCLog::LEVELDB, BCLog::Level::Debug, "%s\n", base);
```
[the remark this answers is not in the document]
Comment (fanquake) [reply src/dbwrapper.cpp:79]: Addressed this.

**Comment by fanquake** [comment]
Moving back to draft while we follow up with more suggestions.

File: contrib/devtools/bitcoin-tidy/example_logprintf.cpp (hunk 9 above)
[the remark this answers is not in the document]
Comment (theuni) [reply contrib/devtools/bitcoin-tidy/example_logprintf.cpp:5]: FWIW, I made this non-default because I don't think it's a good idea for a vanilla `make` to spew warnings _on purpose_. It's bound to set off c-i somewhere eventually.

As @MarcoFalke points out though, we do want it run when we're in control of the environment and not worried about those warnings. In order to run them, I suggest:
```diff
diff --git a/ci/test/06_script_b.sh b/ci/test/06_script_b.sh
index 75d5469267..02358db789 100755
--- a/ci/test/06_script_b.sh
+++ b/ci/test/06_script_b.sh
@@ -150,6 +150,7 @@ fi
 if [ "${RUN_TIDY}" = "true" ]; then
   cmake -B /tidy-build -DLLVM_DIR=/usr/lib/llvm-16/cmake -DCMAKE_BUILD_TYPE=Release -S "${BASE_ROOT_DIR}"/contrib/devtools/bitcoin-tidy
   cmake --build /tidy-build "$MAKEJOBS"
+  cmake --build /tidy-build --target bitcoin-tidy-tests "$MAKEJOBS"

   set -eo pipefail
   cd "${BASE_BUILD_DIR}/bitcoin-$HOST/src/"
```

**Comment by theuni** [comment]
@fanquake Would you mind removing the single use of `NOLINT` and pushing, just so that we can make sure the c-i still fails as expected after all the changes?

Assuming it does, LGTM either with or without addressing [this comment](https://github.com/bitcoin/bitcoin/pull/26296#discussion_r1279354069).

File: contrib/devtools/bitcoin-tidy/example_logprintf.cpp (hunk 9 above)
[the remark this answers is not in the document]
Comment (fanquake) [reply contrib/devtools/bitcoin-tidy/example_logprintf.cpp:5]: Added in the current push.

**Comment by fanquake** [comment]
> Would you mind removing the single use of NOLINT and pushing, just so that we can make sure the c-i still fails as expected after all the changes?

Done.

**Comment by fanquake** [comment]
From: https://cirrus-ci.com/task/6097246463197184

```bash
[100%] Building CXX object CMakeFiles/bitcoin-tidy-tests.dir/example_logprintf.cpp.o
/tmp/cirrus-ci-build/contrib/devtools/bitcoin-tidy/example_logprintf.cpp:65:15: warning: Unterminated format string used with LogPrintf [bitcoin-unterminated-logprintf]
    LogPrintf("hello world!");
              ^
                           \n
/tmp/cirrus-ci-build/contrib/devtools/bitcoin-tidy/example_logprintf.cpp:22:68: note: expanded from macro 'LogPrintf'
#define LogPrintf(...) LogPrintLevel_(LogFlags::NONE, Level::None, __VA_ARGS__)
                                                                   ^
/tmp/cirrus-ci-build/contrib/devtools/bitcoin-tidy/example_logprintf.cpp:21:104: note: expanded from macro 'LogPrintLevel_'
#define LogPrintLevel_(category, level, ...) LogPrintf_(__func__, __FILE__, __LINE__, category, level, __VA_ARGS__)
                                                                                                       ^
/tmp/cirrus-ci-build/contrib/devtools/bitcoin-tidy/example_logprintf.cpp:69:15: warning: Unterminated format string used with LogPrintf [bitcoin-unterminated-logprintf]
    LogPrintf("");
              ^
               \n
/tmp/cirrus-ci-build/contrib/devtools/bitcoin-tidy/example_logprintf.cpp:22:68: note: expanded from macro 'LogPrintf'
#define LogPrintf(...) LogPrintLevel_(LogFlags::NONE, Level::None, __VA_ARGS__)
                                                                   ^
/tmp/cirrus-ci-build/contrib/devtools/bitcoin-tidy/example_logprintf.cpp:21:104: note: expanded from macro 'LogPrintLevel_'
#define LogPrintLevel_(category, level, ...) LogPrintf_(__func__, __FILE__, __LINE__, category, level, __VA_ARGS__)
                                                                                                       ^
/tmp/cirrus-ci-build/contrib/devtools/bitcoin-tidy/example_logprintf.cpp:74:15: warning: Unterminated format string used with LogPrintf [bitcoin-unterminated-logprintf]
    LogPrintf("hello world!...");
              ^
                              \n
/tmp/cirrus-ci-build/contrib/devtools/bitcoin-tidy/example_logprintf.cpp:22:68: note: expanded from macro 'LogPrintf'
#define LogPrintf(...) LogPrintLevel_(LogFlags::NONE, Level::None, __VA_ARGS__)
                                                                   ^
/tmp/cirrus-ci-build/contrib/devtools/bitcoin-tidy/example_logprintf.cpp:21:104: note: expanded from macro 'LogPrintLevel_'
#define LogPrintLevel_(category, level, ...) LogPrintf_(__func__, __FILE__, __LINE__, category, level, __VA_ARGS__)
                                                                                                       ^
[100%] Built target bitcoin-tidy-tests
```

```bash
clang-tidy-16 -p=/tmp/cirrus-ci-build/ci/scratch/build/bitcoin-x86_64-pc-linux-gnu -quiet -load=/tidy-build/libbitcoin-tidy.so /tmp/cirrus-ci-build/ci/scratch/build/bitcoin-x86_64-pc-linux-gnu/src/dbwrapper.cpp
dbwrapper.cpp:79:68: error: Unterminated format string used with LogPrintf [bitcoin-unterminated-logprintf,-warnings-as-errors]
                LogPrintLevel(BCLog::LEVELDB, BCLog::Level::Debug, "%s", base);
                                                                   ^
                                                                      \n
./logging.h:256:45: note: expanded from macro 'LogPrintLevel'
            LogPrintLevel_(category, level, __VA_ARGS__); \
--
```

**Comment by theuni** [comment]
> From: https://cirrus-ci.com/task/6097246463197184
> ...

Nice, thanks!

**Comment by fanquake** [comment]
Dropped the test commit back out.

**Review by theuni — APPROVED** [review]
ACK 1c976c691cc4b20f43071aabf36c7afed1571057

**Review by maflcko — APPROVED** [review]
re-ACK 1c976c691cc4b20f43071aabf36c7afed1571057  👠

<details><summary>Show signature</summary>

Signature:

```
untrusted comment: signature from minisign secret key on empty file; verify via: minisign -Vm "${path_to_any_empty_file}" -P RWTRmVTMeKV5noAMqVlsMugDDCyyTSbA3Re5AkUrhvLVln0tSaFWglOw -x "${path_to_this_whole_four_line_signature_blob}"
RUTRmVTMeKV5npGrKx1nqXCw5zeVHdtdYURB/KlyA/LMFgpNCs+SkW9a8N95d+U4AP1RJMi+krxU1A3Yux4bpwZNLvVBKy0wLgM=
trusted comment: re-ACK 1c976c691cc4b20f43071aabf36c7afed1571057  👠
AUAtgJBmTgwGLADs8IgqfpOQg2nQfQ2+D/7P8lwnKkrIGJwTNAjXOYnonfN9oAT1L5+CTeUJzKaPpGHVi5t7Cw==
```

</details>

**Comment by TheCharlatan** [comment]
ACK 1c976c691cc4b20f43071aabf36c7afed1571057

**Comment by fanquake** [comment]
>  is anything left to be done here 

I missed that Cory had already ACK'd, and was waiting for him to re-ACK. Given that's already happened, I don't think there's anything left here, and docs can be fixed in the followup.

File: contrib/devtools/bitcoin-tidy/logprintf.cpp [hunk 11]
```diff
@@ -0,0 +1,62 @@
...
+
+    /*
+      CWallet wallet;
+      auto walletptr = &wallet;
+      wallet.WalletLogPrintf("foo");
+      wallet->WalletLogPrintf("foo");
+    */
+    finder->addMatcher(
+        cxxMemberCallExpr(
+            thisPointerType(qualType(hasDeclaration(cxxRecordDecl(hasName("CWallet"))))),
```
[the remark this answers is not in the document]
Comment (theuni) [reply contrib/devtools/bitcoin-tidy/logprintf.cpp:46]: Ugh, I was inclined to argue with you because I think the tests should be as tight as possible. But realistically, it's more likely for a class to be forgotten in the checks (as has happened here) than a false-positive in some future class. So I begrudgingly agree.

File: contrib/devtools/bitcoin-tidy/README [hunk 12]
```diff
@@ -0,0 +1,8 @@
+# Bitcoin Tidy
+
+Example Usage:
+
+```bash
+cmake -S . -B build -DLLVM_DIR=/path/to/lib/cmake/llvm -DCMAKE_BUILD_TYPE=Release
```
[the remark this answers is not in the document]
Comment (fanquake) [reply contrib/devtools/bitcoin-tidy/README:6]: Done in #28245.

**Comment by theuni** [comment]
Yep, will do. I was just waiting for the others to settle.

File: contrib/devtools/bitcoin-tidy/CMakeLists.txt [hunk 13]
```diff
@@ -0,0 +1,45 @@
...
+
+project(bitcoin-tidy VERSION 1.0.0 DESCRIPTION "clang-tidy checks for Bitcoin Core")
+
+include(GNUInstallDirs)
+
+set(CMAKE_CXX_STANDARD 17)
+set(CMAKE_CXX_STANDARD_REQUIRED True)
+set(CMAKE_CXX_EXTENSIONS False)
+
+# TODO: Figure out how to avoid the terminfo check
```
[the remark this answers is not in the document]
Comment (theuni) [reply contrib/devtools/bitcoin-tidy/CMakeLists.txt:11]: Actually, looking around, I think it makes more sense to just delete this comment. Autotools checks for a bazillion things it doesn't need. This is just one of those quirks that can't be turned off.

(The issue is that LLVM's CMake file checks for terminfo even though we don't need it. It doesn't seem to be a problem in the real world. If we encounter an actual issue, THEN we can worry about working around it.)
